stdlib/universe/map2.go

```diff
@@ -0,0 +1,276 @@
+package universe
+
+import (
+	"context"
+	"sort"
+
+	"github.com/apache/arrow/go/arrow/memory"
+	"github.com/influxdata/flux"
+	"github.com/influxdata/flux/array"
+	"github.com/influxdata/flux/arrow"
+	"github.com/influxdata/flux/codes"
+	"github.com/influxdata/flux/compiler"
+	"github.com/influxdata/flux/execute"
+	"github.com/influxdata/flux/execute/table"
+	"github.com/influxdata/flux/internal/arrowutil"
+	"github.com/influxdata/flux/internal/errors"
+	"github.com/influxdata/flux/internal/mutable"
+	"github.com/influxdata/flux/values"
+)
+
+//go:generate -command tmpl ../../gotool.sh github.com/benbjohnson/tmpl
+//go:generate tmpl -data=@../../array/types.tmpldata -o map2.gen.go map2.gen.go.tmpl
+
+type mapTransformation2 struct {
+	ctx context.Context
+	fn  *execute.RowMapFn
+}
+
+func newMapTransformation2(ctx context.Context, id execute.DatasetID, spec *MapProcedureSpec, mem memory.Allocator) (execute.Transformation, execute.Dataset, error) {
+	tr := &mapTransformation2{
+		ctx: ctx,
+		fn:  execute.NewRowMapFn(spec.Fn.Fn, compiler.ToScope(spec.Fn.Scope)),
+	}
+	return execute.NewGroupTransformation(id, tr, mem)
+}
+
+func (m *mapTransformation2) Process(chunk table.Chunk, d *execute.TransportDataset, mem memory.Allocator) error {
+	// The current version of map just silently drops
+	// empty tables so let's just do that.
+	if chunk.Len() == 0 {
+		return nil
+	}
+
+	// Prepare the compiled function for the set of columns.
+	cols := chunk.Cols()
+	fn, err := m.fn.Prepare(cols)
+	if err != nil {
+		return err
+	}
+
+	// Execute function.
+	cols, arrs, err := m.execute(chunk, fn, mem)
+	if err != nil {
+		return err
+	}
+	return m.regroup(cols, chunk.Key(), arrs, d, mem)
+}
+
+// initialize will initialize the builders using the column metadata.
+func (m *mapTransformation2) initialize(cols []flux.ColMeta, mem memory.Allocator) []array.Builder {
+	builders := make([]array.Builder, len(cols))
+	for i, col := range cols {
+		builders[i] = arrow.NewBuilder(col.Type, mem)
+	}
+	return builders
+}
+
+// execute will execute the function on each row of the table chunk.
+// It will return the new columns after the function was executed.
+// These columns are not organized by group keys and may require regrouping
+// if a group key column no longer contains consistent values.
+func (m *mapTransformation2) execute(chunk table.Chunk, fn *execute.RowMapPreparedFn, mem memory.Allocator) ([]flux.ColMeta, []array.Interface, error) {
```

Comment:
This should be the only function that has to get modified for the vectorized version. The return value is exactly what we need and the parts around this all rely on the same output, but this function would need to be updated to also execute the vectorized version.

Reply:
So if the function that we have here has a type like `(r: {a: v[int], b: v[int]}) => {c: v[int]}`, could we expect that the above code in `createSchema` will just work?

Also, can `createSchema` be a stand-alone function rather than a method?

Reply:
It kind of depends on the type that gets returned. We may have to teach create schema to look at the element type if the function is vectorized.

It would be possible for `createSchema` to be a standalone function, but I usually try to scope these by making them method calls so I don't have to worry about name conflicts and using overly generic names.